### PR TITLE
XCM Pallet

### PR DESCRIPTION
**Pull request by alistair-singh**
Resolves: SNO-228, SNO-227

## Summary

1. Introduce an xcm pallet to emit events for transfer failures.

Success:
```json
{
   "method": "TransferSent",
   "section": "xcmSupport",
   "index": "0x1500",
   "data": [
     {
       "assetId": 0,
       "sender": "0x89b4ab1ef20763630df9743acf155865600daff2",
       "recipient": "0xd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d",
       "amount": 100000000000000,
       "paraId": 1001,
       "fee": 4000000,
     }
   ]
 }
```

Error:
```json
{
    "method": "TransferFailed",
    "section": "xcmSupport",
    "index": "0x1501",
    "data": {
      "info": {
        "assetId": 0,
        "sender": "0x89b4ab1ef20763630df9743acf155865600daff2",
        "recipient": "0xd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d",
        "amount": 100000000000000,
        "paraId": 2001,
        "fee": 4000000,
      },
      "error": {
        "Module": {
          "index": 21,
          "error": "0x02000000"
        }
      }
   }
}
```

2. Removed duplicated transaction handling.
3. Changed the `reserve_transfer` trait to not return `DispatchResult`.
4. Added pallet errors.
5. Moved all types into `xcm-support-primitives` crate.
6. Updated metadata to include the new events and types.

E2E tests pass.

## Todo
- [x] In the E2E stack test that these events are emitted.

### CODE REVIEW

File: parachain/pallets/eth-app/src/lib.rs [hunk 1]
```diff
@@ -152,27 +149,15 @@ pub mod pallet {
...
-						log::error!(
-							"Failed to execute xcm transfer to parachain {} - {:?}.",
-							destination.para_id,
-							err
-						);
-						return TransactionOutcome::Rollback(DispatchError::Other("foo").into());
-					}
-					TransactionOutcome::Commit(Ok(()))
-				});
+				T::XcmReserveTransfer::reserve_transfer(0, sender, &recipient, amount, destination);
```
Comment (claravanstaden) [inline parachain/pallets/eth-app/src/lib.rs:160]: Clean! ✨

File: parachain/pallets/xcm-support/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+sp-runtime = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.19" }
+sp-io = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.19" }
+
+[features]
+default = [ "std" ]
+std = [
+	"codec/std",
+	"scale-info/std",
+	"frame-support/std",
+	"frame-system/std",
```
Comment (claravanstaden) [inline parachain/pallets/xcm-support/Cargo.toml:41]: I think sp-core is missing here:

```suggestion
	"frame-system/std",
	"sp-core/std",
```

File: parachain/pallets/xcm-support/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+xcm = { git = "https://github.com/paritytech/polkadot.git", branch = "release-v0.9.19", default-features = false }
+xcm-executor = { git = "https://github.com/paritytech/polkadot.git", branch = "release-v0.9.19", default-features = false }
+xcm-builder = { git = "https://github.com/paritytech/polkadot.git", branch = "release-v0.9.19", default-features = false }
+
+snowbridge-xcm-support-primitives = { path = "../../primitives/xcm-support", default-features = false }
+
+[dev-dependencies]
+serde = { version = "1.0.136" }
+sp-core = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.19" }
+sp-runtime = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.19" }
```
Comment (claravanstaden) [inline parachain/pallets/xcm-support/Cargo.toml:32]: Do these two dependencies need to be here if they are already in `[dependencies]`?

**Review by claravanstaden — APPROVED** [review]
Very nice! 👏

File: parachain/pallets/xcm-support/Cargo.toml (hunk 3 above)
Comment (alistair-singh) [reply parachain/pallets/xcm-support/Cargo.toml:32]: I am unsure. Will remove them and see if it compiles.

Comment (alistair-singh) [reply parachain/pallets/xcm-support/Cargo.toml]: Ok removed these two dev-dependencies. Thanks!

**Review by vgeddes — APPROVED** [review]
Looks good!

**Pull request merged**